platform/linux-generic/arch/x86/odp/api/abi/random.h

```diff
@@ -0,0 +1,107 @@
+/* Copyright (c) 2021, Nokia
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+/*
+ * These functions implement ODP_RANDOM_TRUE type random data on x86 using
+ * rdrand and rdseed [1] via compiler builtin functions.
+ *
+ * We use rdseed if available. This is the "truest" random source on x86, but it
+ * may have low throughput. Otherwise we use rdrand.
+ *
+ * Note that there may be issues with the quality or security of rdrand and
+ * rdseed. [2]
+ *
+ * [1] Intel Digital Random Number Generator (DRNG) Software Implementation
+ *     Guide. John P Mechalas, 17 October 2018.
+ *     https://www.intel.com/content/www/us/en/developer/articles/guide/intel-digital-random-number-generator-drng-software-implementation-guide.html
+ *
+ * [2] RDRAND. Wikipedia, 29 September 2021.
+ *     https://en.wikipedia.org/wiki/RDRAND#Reception
+ */
+
+#ifndef ODP_API_ABI_RANDOM_H_
+#define ODP_API_ABI_RANDOM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <odp/api/spec/random.h>
+
+#include <stdint.h>
+
+int _odp_random_have_true_generic(void);
+int32_t _odp_random_true_data_generic(uint8_t *buf, uint32_t len);
+
+#ifdef __RDRND__
+
+static inline int _odp_random_have_true(void)
+{
+	return 1;
+}
+
+#ifdef __RDSEED__
+#define _rd64 __builtin_ia32_rdseed_di_step
+#define _rd32 __builtin_ia32_rdseed_si_step
+#define _rd16 __builtin_ia32_rdseed_hi_step
+#else
+#define _rd64 __builtin_ia32_rdrand64_step
```

Comment:
To me rdrand does not appear suitable for ODP_RANDOM_TRUE. The Intel document describes how the numbers returned by rdseed have "multiplicative brute-force prediction resistance", which is what I would expect from ODP_RANDOM_TRUE. The document also describes (section 5.2.6) how to derive seeds from rdrand output, which is something that the user of ODP_RANDOM_TRUE data should be be expected to do (as opposed to just using the data directly).

So, I would provide ODP_RANDOM_TRUE only if rdseed is available and use rdrand for ODP_RANDOM_CRYPTO.

Reply:
Fixed in v3.